### PR TITLE
[Robust Shadow 2]Add few functions for robust shadow estimation

### PR DESCRIPTION
**Pull request by Min-Li**
## Description

Add few functions for robust shadow estimation

### License

- [X] I license this contribution under the terms of the GNU GPL, version 3 and grant Unitary Fund the right to provide additional permissions as described in section 7 of the GNU GPL, version 3.

Before opening the PR, please ensure you have completed the following where appropriate.
- [X] I added unit tests for new code.
- [X] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [X] I used [Google-style](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods) docstrings for functions.
- [X] I [updated the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md) where relevant.
- [ ] Added myself / the copyright holder to the AUTHORS file

### CODE REVIEW

File: mitiq/shadows/shadows_utils.py [hunk 1]
```diff
@@ -8,13 +8,74 @@
...
 import mitiq
 
+from itertools import product
+
+Paulis = [
+    cirq.I._unitary_(),
+    cirq.X._unitary_(),
+    cirq.Y._unitary_(),
+    cirq.Z._unitary_(),
+]
```
Comment (natestemen) [inline mitiq/shadows/shadows_utils.py:26]: Globals are usually all caps.
```suggestion
PAULIS = [
    cirq.I._unitary_(),
    cirq.X._unitary_(),
    cirq.Y._unitary_(),
    cirq.Z._unitary_(),
]
```

File: mitiq/shadows/shadows_utils.py [hunk 2]
```diff
@@ -8,13 +8,74 @@
...
-from typing import List, Tuple
+from typing import Tuple, List, Any
 
 import numpy as np
 from numpy.typing import NDArray
+import cirq
 
 import mitiq
 
+from itertools import product
```
Comment (natestemen) [inline mitiq/shadows/shadows_utils.py:19]: Remember to group imports as suggested in https://github.com/unitaryfund/mitiq/pull/1907#discussion_r1275578388.

File: mitiq/shadows/shadows_utils.py [hunk 3]
```diff
@@ -8,13 +8,74 @@
...
+
+
+def kronecker_product(matrices: List[NDArray[Any]]) -> NDArray[Any]:
+    """
+    Returns the Kronecker product of a list of matrices.
+    """
+    result = matrices[0]
+    for matrix in matrices[1:]:
+        result = np.kron(result, matrix)
+    return result
```
Comment (natestemen) [inline mitiq/shadows/shadows_utils.py:36]: Is this more or less readable? I slightly favor `reduce`, but up to you.
```suggestion
    return functools.reduce(np.kron, matrices)
```

File: mitiq/shadows/shadows_utils.py [hunk 4]
```diff
@@ -8,13 +8,74 @@
...
+
+
+# Input must be a 2x2 matrix, output is a vector of 4 elements
```
Comment (natestemen) [inline mitiq/shadows/shadows_utils.py:39]: This comment should go in the docstring.

File: mitiq/shadows/shadows_utils.py [hunk 5]
```diff
@@ -8,13 +8,74 @@
...
+def operator_ptm_vector_rep(opt: NDArray[Any]) -> NDArray[Any]:
+    r"""
+    Returns the PTM vector representation
+    :math:`|o\rangle\!\rangle\in \mathcal{H}_{4^n}`
+    of an operator :math:`opt\in \mathcal{L}(\mathcal{H}_{2^n})`.
```
Comment (natestemen) [inline mitiq/shadows/shadows_utils.py:44]: You'll have to double check this renders correctly, but $opt\in \mathcal{L}(\mathcal{H}_{2^n})$ is not what we want.
```suggestion
    of an operator :math:`\mathtt{opt}\in \mathcal{L}(\mathcal{H}_{2^n})`.
```

File: mitiq/shadows/shadows_utils.py [hunk 6]
```diff
@@ -8,13 +8,74 @@
...
+    """
+    # vector i-th entry is math:`d^{-1/2}Tr(oP_i)`
+    # where P_i is the i-th Pauli matrix
+    assert (
+        len(opt.shape) == 2 and opt.shape[0] == opt.shape[1]
+    ), "Input must be a square matrix"
```
Comment (natestemen) [inline mitiq/shadows/shadows_utils.py:50]: This should be a `TypeError`, right?

File: mitiq/shadows/test/test_shadows_utils.py [hunk 7]
```diff
@@ -4,16 +4,57 @@
...
+def test_eigenvalues_to_bitstring():
+    values = [-1, 1, 1]
+    assert eigenvalues_to_bitstring(values) == "100"
+
+
+def test_bitstring_to_eigenvalues():
+    bitstring = "100"
+    np.testing.assert_array_equal(
+        bitstring_to_eigenvalues(bitstring), np.array([-1, 1, 1])
+    )
```
Comment (natestemen) [inline mitiq/shadows/test/test_shadows_utils.py:49]: Let's add a test to ensure these two functions are inverses of each other.

File: mitiq/shadows/shadows_utils.py [hunk 8]
```diff
@@ -8,13 +8,74 @@
...
+    """Converts eigenvalues to bitstring. e.g., [-1,1,1] -> '100'"""
+    return "".join(["1" if v == -1 else "0" for v in values])
+
+
+def bitstring_to_eigenvalues(bitstring: str) -> List[int]:
+    """Converts bitstring to eigenvalues. e.g., '100' -> [-1,1,1]"""
+    return [1 if b == "0" else -1 for b in bitstring]
+
+
+def create_string(n: int, loc_list: List[int]) -> str:
```
Comment (natestemen) [inline mitiq/shadows/shadows_utils.py:71]: Let's avoid single letter argument variable names. Maybe `string_length`? In a similar vein, what does `loc_list` mean? Location list? Can we make this more explicit?

File: mitiq/shadows/shadows_utils.py [hunk 9]
```diff
@@ -8,13 +8,74 @@
...
+    """
+    This function returns a string of length n with 1s at the locations
+    specified by loc_list and 0s elsewhere.
+    """
+    loc_set = set(loc_list)  # Convert list to set for efficient lookups
```
Comment (natestemen) [inline mitiq/shadows/shadows_utils.py:76]: Should we just take in a `set` directly?

File: mitiq/shadows/shadows_utils.py [hunk 10]
```diff
@@ -36,10 +97,8 @@ def local_clifford_shadow_norm(obs: mitiq.PauliString) -> float:
     """
     Calculate shadow norm of an operator with random unitary sampled from local
     Clifford group.
-
     Args:
-        obs: An observable in terms of a ``mitiq.PauliString`` object.
-
+        opt: a self-adjoint operator.
```
Comment (natestemen) [inline mitiq/shadows/shadows_utils.py:101]: Is this change intended? I added a commit (https://github.com/unitaryfund/mitiq/pull/1907/commits/a6c579348fe6c55b0291dce80a3012d0106c64a2) in https://github.com/unitaryfund/mitiq/pull/1907/ to fix this.

**Review by natestemen — CHANGES_REQUESTED** [review]

**Review by andreamari — COMMENTED** [review]
From my side, just a comment about a function name and its docstring.

File: mitiq/shadows/shadows_utils.py [hunk 11]
```diff
@@ -88,16 +146,28 @@ def n_measurements_opts_expectation_bound(
...
     """
-    Calculate the fidelity between a state vector and a density matrix.
+    fidelity is a measure of the "closeness" of two quantum states.
+    It expresses the probability that one state will pass a test to
+    identify as the other.
     Args:
-        state_vector: The vector whose norm we want to calculate.
-        rho: The operator whose norm we want to calculate.
+        sigma: Quantum state.
+        rho: Quantum state
```
Comment (andreamari) [inline mitiq/shadows/shadows_utils.py:158]: to check

File: mitiq/shadows/shadows_utils.py [hunk 12]
```diff
@@ -8,13 +8,73 @@
...
+    r"""
+    Returns the PTM vector representation
+    :math:`|o\rangle\!\rangle\in \mathcal{H}_{4^n}`
+    of an operator :math:`o\in \mathcal{L}(\mathcal{H}_{2^n})`.
+    """
+    # vector i-th entry is math:`d^{-1/2}Tr(oP_i)`
+    # where P_i is the i-th Pauli matrix
+    assert (
+        len(obs.shape) == 2 and obs.shape[0] == obs.shape[1]
+    ), "Input must be a square matrix"
```
Comment (andreamari) [inline mitiq/shadows/shadows_utils.py:49]: to check. Seems a reverted change.

Comment (Min-Li) [reply mitiq/shadows/shadows_utils.py]: thanks, have changed

File: mitiq/shadows/shadows_utils.py (hunk 5 above)
Comment (Min-Li) [reply mitiq/shadows/shadows_utils.py]: Thanks, will change it fonts

File: mitiq/shadows/shadows_utils.py (hunk 11 above)
Comment (Min-Li) [reply mitiq/shadows/shadows_utils.py]: Oh, Andrea, here $\sigma$ and $\rho$  might not be a quantum state... I left some comments for why it is on #1942 under your comment, please check whether you are happy with my comments.  I have change it to
```
sigma: A state in terms of square matrix or vector.
rho: A state in terms square matrix or vector.
```

File: mitiq/shadows/shadows_utils.py (hunk 9 above)
Comment (Min-Li) [reply mitiq/shadows/shadows_utils.py]: this `set` operation just wants to ensure there are no repetitive elements in the `loc_list`.

File: mitiq/shadows/shadows_utils.py (hunk 11 above)
Comment (natestemen) [reply mitiq/shadows/shadows_utils.py]: @Min-Li please link directly to your comment regarding this instead of the PR.

File: mitiq/shadows/shadows_utils.py [hunk 13]
```diff
@@ -8,13 +8,74 @@
...
+    result = matrices[0]
+    for matrix in matrices[1:]:
+        result = np.kron(result, matrix)
+    return result
+
+
+# Input must be a 2x2 matrix, output is a vector of 4 elements
+def operator_ptm_vector_rep(opt: NDArray[Any]) -> NDArray[Any]:
+    r"""
+    Returns the PTM vector representation
```
[the remark this answers is not in the document]
Comment (andreamari) [reply mitiq/shadows/shadows_utils.py]: ok sounds good.

File: mitiq/shadows/shadows_utils.py (hunk 11 above)
Comment (andreamari) [reply mitiq/shadows/shadows_utils.py]: Directly linking the comment is always better. In any case, the current docstring for sigma end rho seems good to me.

**Review by andreamari — APPROVED** [review]

**Review by natestemen — APPROVED** [review]
